fix(gossipsub): prevent erroneously duplicate message IDs

dariusc93 opened the pull request:
## Description

Previously, we only mutably borrowed the `last_seq_no` in the current scope but did not modify the underlying number. This is because `u64` is copy and calling `wrapping_add` consumes `self` so the compiler just copied it. We introduce a new-type instead that is not `Copy`.

Additionally, `wrapping_add` and initializing with a random u64 might actually warp the number and thus not give us sequential numbers as intended in #3551. To solve this, we initialize with the current unix timestamp in nanoseconds. This allows a node to publish 1000000 messages a second and still not reuse sequence numbers even after a restart / re-initialization of the configuration. This is also what the go implementation does.

Resolves #3714.

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>
  
<!--
Please write a summary of your changes and why you made them.
This section will appear as the commit message after merging.
Please craft it accordingly.
For a quick primer on good commit mesages, check out this blog post: https://cbea.ms/git-commit/

Please include any relevant issues in here, for example:

Related https://github.com/libp2p/rust-libp2p/issues/ABCD.
Fixes https://github.com/libp2p/rust-libp2p/issues/XYZ.
-->

## Notes & open questions

<!--
Any notes, remarks or open questions you have to make about the PR which don't need to go into the final commit message.
-->

## Change checklist

<!-- Please add a Changelog entry in the appropriate crates and bump the crate versions if needed. See <https://github.com/libp2p/rust-libp2p/blob/master/docs/release.md#development-between-releases>-->

- [x] I have performed a self-review of my own code
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] A changelog entry has been made in the appropriate crates

dariusc93 committed:
fix: Mutablely borrow last_seq_no

dariusc93 committed:
chore: Update CHANGELOG.md

thomaseizinger commented on protocols/gossipsub/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.44.3 - unreleased
+- Mutablely borrow last sequential number. See [PR 3716]
```
```suggestion

- Fix erroneously duplicate message IDs. See [PR 3716].
```

thomaseizinger commented on protocols/gossipsub/CHANGELOG.md:
```diff
@@ -1,3 +1,8 @@
+## 0.44.3 - unreleased
```
Can you bump the version in the Cargo.toml too please?

dariusc93 replied:
Done

thomaseizinger reviewed (approved):
I updated the title and PR description, let me know if I understood the problem and fix correctly!

dariusc93 committed:
Update protocols/gossipsub/CHANGELOG.md

Co-authored-by: Thomas Eizinger <thomas@eizinger.io>

dariusc93 committed:
Update Cargo.toml

dariusc93 committed:
chore: Update Cargo.lock

dariusc93 committed:
Merge branch 'master' into seq-mut-fix

thomaseizinger committed:
Don't make `PublishConfig` clonable

We should never need to clone this, it might break the uniqueness
guarantee of sequence numbers.

thomaseizinger committed:
Wrap sequence number in new type

Basic data type are easily handled the wrong way because they implement
Copy.

This newtype doesn't implement Copy and increments every time to access
the sequence number.

thomaseizinger committed:
Use milliseconds

thomaseizinger committed:
Switch to nano seconds

AgeManning reviewed (approved):
Yep this makes sense to me, the panic at u64::max() is for someone living 600k years in the future's problem (if they can keep a libp2p node running that long without updating :).) 

This also doesn't effect us as we don't use sequence numbers, so all good from our end, thanks for the ping :).

mergify[bot] committed:
Merge branch 'master' into seq-mut-fix
